proposals/code-signing-for-homebrew.md

```diff
@@ -0,0 +1,101 @@
+
+[Sigstore](https://www.sigstore.dev/) is a Linux Foundation project aimed at making codesigning available to everybody. It uses a combination of open standards (OpenID Connect, WebPKI, Certificate Transparency) to bind non-cryptographic identities (such as emails or GitHub Actions workflow identities) to short-lived signing keys, which can then be used to sign for artifacts much like `gpg` or any other signing tool.
+
+## Technical Items
+
+### Sigstore signatures for official Homebrew taps
+
+Homebrew's packaging processes are built around GitHub Actions. In particular, Homebrew uses GitHub Actions to perform builds of all core formulae (package specifications, like LLVM, `bash`, or `git`) into bottles (binary distributions that can be installed on client machines, similar to an `.app` or `.pkg`). Bottles are then hosted on GitHub Packages (or other hosts, for third-party taps) and are installed by end-user clients e.g. during `brew install` invocations.
+
+Because Homebrew uses GitHub Actions for all `homebrew-core` builds, the Homebrew CI has trivial access to GitHub Actions' [ambient OpenID Connect identities](https://docs.github.com/en/actions/deployment/security-hardening-your-deployments/about-security-hardening-with-openid-connect). These identities can in turn be used to sign with Sigstore, meaning that Homebrew's existing CI workflows would require relatively little additional configuration (and zero external key management) to produce valid Sigstore signatures for each bottle build.
+
+The high-level flow for bottle signing:
+
+1. Existing bottle-building workflows will gain a signing step, which will only run as a finalization step for bottles that appear on `master` (i.e., not just the temporary bottles that are built as part of the normal PR lifecycle).
+2. The signing step is responsible for producing a Sigstore signature that _attests_ to the bottle. This should include, at minimum, the bottle's SHA256 digest and the bottle's filename. The digest is necessary to produce a proof that the signature is for a particular bottle's contents, and the filename is necessary to produce a proof that the signature is intended to accompany a particular install step (i.e. `brew install foo` should not verify if an attacker substitutes `foo`'s bottle for a correctly signed bottle of bar, or even a different version of foo than the resolver produces).
+3. The signing step is responsible for uploading the resulting signature to the same object storage as the bottle, and ensuring that the uploaded signature is addressable in a consistent and publicly derivable manner from just the formula and bottle's own metadata. In other words: a client that knows how to access the bottle's resource should also be able to access the bottle's signature resource with no additional metadata. 
```

Comment:
Yes! I actually wasn't aware of `cosign triangulate`, but this looks like it will be a helpful reference for our discovery step here (esp. if we end up using a dedicated attestation storage API as mentioned by @steiza)

Reply:
84124b2410d6074d9e8360ba35b402839bc8c898